shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java

```diff
@@ -133,6 +135,24 @@ public boolean setComposingText(CharSequence text, int newCursorPosition) {
         return result;
     }
 
+    @Override
+    public boolean finishComposingText() {
+        boolean result = super.finishComposingText();
+
+        if (Build.VERSION.SDK_INT >= 21) {
+            // Update the keyboard with a reset/empty composing region. Critical on
+            // Samsung keyboards to prevent punctuation duplication.
+            CursorAnchorInfo.Builder builder = new CursorAnchorInfo.Builder();
+            builder.setComposingText(-1, "");
+            CursorAnchorInfo anchorInfo = builder.build();
+            mImm.updateCursorAnchorInfo(mFlutterView, anchorInfo);
+        }
```

Comment:
Is this expensive? Should we be limiting it to only when a samsung keyboard is in use (or perhaps only on samsung devices)?

Reply:
I can gate this behind a Samsung check, but I wanted to avoid adding device/OEM specific checks as it makes the code much messier. I don't think this is particularly expensive, and it is only performed occasionally as it only runs when manually triggered by keyboard.

Reply:
SGTM.